### ISSUE TITLE
Issue in JSON manifest

### ISSUE THREAD

**Issue by Priyaaxle** [usnistgov/WIPP-frontend#161]
**Summary**

When we install json manifest while creating the plugin, if there is any issue in the json file, we could not add task in the workflow. 

Below is the json file used for creating the plugin.

{
    "name": "Feature Extraction",
    "version": "0.2.0",
    "title": "Feature Extraction",
    "description": "Extract shape and intensity features from images.",
    "repository": "https://github.com/labshare/polus-plugins",
    "website": "https://ncats.nih.gov/preclinical/core/informatics",
    "citation": "",
    "containerId": "labshare/polus-feature-extraction-plugin:0.2.0",
    "inputs": [
      {
        "name": "angleStart",
        "type": "integer",
        "description": "Angle start degree to calculate feret diameter",
        "required": false
      },
      {
        "name": "angleStop",
        "type": "integer",
        "description": "Angle end degree to calculate feret diameter",
        "required": false
      },
      {
        "name": "boxSize",
        "type": "integer",
        "description": "Boxsize to calculate feret diameter",
        "required": false
      },
      {
        "name": "csvfile",
        "type": "array",
        "description": "Select option to save csv file separately for each image or one csv for all images",
        "required": true
      },
      {
        "name": "features",
        "type": "array",
        "description": "Select required features",
        "required": true
      },
      {
        "name": "inpDir",
        "type": "collection",
        "description": "Input image collection to be processed by this plugin",
        "required": true
      },
      {
        "name": "pixelDistance",
        "type": "integer",
        "description": "Pixel distance to calculate the neighbors touching cells",
        "required": false
      },
      {
        "name": "segDir",
        "type": "collection",
        "description": "Input image collection containing image segmentations",
        "required": false
      }
    ],
    "outputs": [
      {
        "name": "outDir",
        "type": "collection",
        "description": "Output collection"
      }
    ],
    "ui": [
      {
        "key": "inputs.angleStart",
        "title": "Input value",
        "description": "Angle start degree to calculate feret diameter"
      },
      {
        "key": "inputs.angleStop",
        "title": "Input value",
        "description": "Angle end degree to calculate feret diameter"
      },
      {
        "key": "inputs.boxSize",
        "title": "Input value",
        "description": "Boxsize to calculate feret diameter"
      },
      {
        "key": "inputs.csvfile",
        "title": "Output csv file",
        "description": "Select option to save csv file separately for each image or one csv for all images"
      },
      {
        "key": "inputs.features",
        "title": "Select features",
        "description": "Select required features"
      },
      {
        "key": "inputs.inpDir",
        "title": "Input collection",
        "description": "Input image collection to be processed by this plugin"
      },
      {
        "key": "inputs.pixelDistance",
        "title": "Input value",
        "description": "Pixel distance to calculate the neighbors touching cells"
      },
      {
        "key": "inputs.segDir",
        "title": "Input collection",
        "description": "Input image collection containing image segmentations"
      }
    ]
  }

cc @Nicholas-Schaub

**Comment by MyleneSimon** [comment]
Hi @Priyaaxle thank you for reporting the issue. It seems that it is coming from the inputs `csvfile` and `features`.
For now, the type `array` is only supported for creating an array of strings from a pre-set list of available strings (described in `options` -> `items` -> `oneOf` in the manifest).
See for example the `features` input description in the Feature2DJava plugin: [manifest](https://github.com/usnistgov/WIPP/blob/e9fb3d3a4cc507eb39c1864e5e9d4df27d43cad2/plugins/wipp-feature2djava-plugin.json#L33)
```
{
      "name": "features",
      "type": "array",
      "options": {
        "items": { 
            "type": "string", 
            "title": "Select feature",
            "oneOf": 
                [
                    { "description": "Area", "enum": [ "Feature2DJava_Area" ] }, 
                    { "description": "Mean", "enum": [ "Feature2DJava_Mean" ] }, 
                    { "description": "Aspect Ratio BB", "enum": [ "Feature2DJava_Aspect_Ratio_BB" ] }, 
                   ...
                ], 
            "default": "Feature2DJava_Area",
            "widget": "select" ,    
            "minItems": 1, 
            "uniqueItems": "true"
        },
      "required": "true",
      "description": "Features to select"
      }
    }
```

**Comment by Nicholas-Schaub** [comment]
@MyleneSimon Thank you for help debugging the json manifest.

In offline discussions with @Priyaaxle we found an issue that is larger than just the problem with this individual manifest. The problem has to do with Priya's manifest causing the WIPP workflow builder to no longer add tasks to a new workflow. We had this problem in the past when there was a problem in Java 2D feature extraction plugin, where adding that manifest suddenly caused WIPP to no longer permit addition of tasks. It seems like additional validation on json manifests might need to be done, or a bug in WIPP might need to be fixed so that a faulty json manifest does not cause issues for the workflow builder.

**Comment by MyleneSimon** [comment]
@Nicholas-Schaub so do you mean that adding an invalid manifest prevents you from creating tasks using plugins that do have a valid manifest? This is something that was fixed #120, but if this is still happening then we need to look into that again (the fix was probably not handling all possible cases for manifest invalidity). I cannot reproduce the issue with the particular manifest shared here, but we could test with other ones.
I agree that indeed plugin manifest validation during plugin registration is needed, this is on our radar as well.

**Comment by Nicholas-Schaub** [comment]
@MyleneSimon I just tested on my local deployment that uses the latest version in the CI pipeline, and when I try to select the plugin it shows an error (so it appears the problem has already been resolved). This issue occurred on our QA cluster, so it's possible that the fix was not incorporated yet. I think this issue can be closed.

**Comment by Priyaaxle** [comment]
@MyleneSimon Thank you for helping in resolving the issue.

**Comment by MyleneSimon** [comment]
Alright, closing for now then, don't hesitate to reopen if the issue occurs again in the latest versions (this fix was incorporated since the 3.0.0-beta, so best options would be updating QA to the 3.0.0-beta.1 from November 2019 or waiting for the 3.0.0-beta2 later this month to get the fix).
